mlx_graphs/datasets/superpixel.py

```diff
@@ -0,0 +1,214 @@
+    "MNIST": "mnist_75sp",
+    "CIFAR10": "cifar10_150sp",
+}
+
+
+def sigma(distances: mx.array, k: int = 8) -> mx.array:
+    """
+    Computes the scale parameter sigma defined as the averaged distance xk of the k
+    nearest neighbors for each node in the distances matrix.
+
+    See Equation (47) in `<http://arxiv.org/abs/2003.00982>`_ for more details.
+
+    Args:
+        distances:
+        k: nearest neighbors to consider
+
+    Returns:
+        The value of sigma for each node.
+    """
+    num_nodes = distances.shape[0]
+
+    if k > num_nodes:
+        # handle graphs with num_nodes less than k
+        sigma = mx.array([1] * num_nodes).reshape(num_nodes, 1)
+    else:
+        # get knns for each node
+        knns = mx.partition(distances, k, axis=-1)[:, : k + 1]
+        sigma = knns.sum(axis=1).reshape((knns.shape[0], 1)) / k
+
+    return sigma + 1e-8
+
+
+def image_to_superpixel_adjacency_matrix(
+    coordinates: mx.array, features: mx.array, use_features: bool = True
+) -> mx.array:
+    """
+    Computes a k-NN adjacency matrix as in Equation (47) in
+    `<http://arxiv.org/abs/2003.00982>`_.
+
+    Args:
+        coordinates: coordinates of each pixel/node
+        features: features of each pixel/node
+        use_features: whether to use features in computing the adjacency matrix.
+            Defaults to True
+
+    Returns:
+        Adjacency matrix
+    """
+    coord = coordinates.reshape(-1, 2)
+    coord_dist = pairwise_distances(coord, coord)
+
+    if use_features:
+        features_dist = pairwise_distances(features, features)
+        adjacency_matrix = mx.exp(
+            -((coord_dist / sigma(coord_dist)) ** 2)
+            - (features_dist / sigma(features_dist)) ** 2
+        )
+    else:
+        adjacency_matrix = mx.exp(-((coord_dist / sigma(coord_dist)) ** 2))
+
+    # convert to symmetric matrix without self-loops
+    adjacency_matrix = 0.5 * (adjacency_matrix + adjacency_matrix.T)
+    for i in range(adjacency_matrix.shape[0]):
+        adjacency_matrix[i, i] = 0
+    return adjacency_matrix
+
+
+def compute_edges_list(
+    adjacency_matrix: mx.array, k: int = 9
+) -> tuple[mx.array, mx.array]:
+    """
+    Compute list of knn nodes per node (and features)
+
+    Args:
+        adjacency_matrix: the adjacency matrix
+        k: the number of nearest neighbors
+
+    Returns:
+        list of knns for each node and list of their features
+    """
+    num_nodes = adjacency_matrix.shape[0]
+    new_kth = num_nodes - k
+
+    if num_nodes > k:
+        knns = mx.argpartition(adjacency_matrix, new_kth - 1, axis=-1)[:, new_kth:-1]
+        knn_values = mx.partition(adjacency_matrix, new_kth - 1, axis=-1)[:, new_kth:-1]
+    else:
+        # for graphs with less than k nodes the resulting graph will be fully connected.
+        knns = mx.repeat(mx.arange(num_nodes), num_nodes).reshape(num_nodes, num_nodes)
+        knn_values = adjacency_matrix
+
+        # TODO: eventually remove self-loops
```

Comment:
do you need to remove self-loops here?

Reply:
in practice it seems the first part of the if is the that's always executed for this two datasets - so I avoided adding the self-loop removal as it requires numpy for the moment (because we can't index with booleans in mlx). I'll add that part anyways though.